### ISSUE TITLE
MIDI length limits/representation issues?

### ISSUE THREAD

**Issue by gwern** [Rick-McCoy/Reformer-pytorch#2]
Does this Reformer truncate or hide almost all of MIDI files during training?

I was [generating MIDI files as represented in ABC](https://www.gwern.net/GPT-2-music#generating-midi-with-30k-context-windows) using GPT-2 and thought I'd compare with Reformer. Since this implementation had MIDI support it was convenient to get working (although I found that the version dependencies on Pytorch omit that you need a version recent enough to have that Tensorboard compatibility module, lengths appear to only be powers-of-two, and also the code assumes you've already created a `./samples/` directory & crashes if you don't manually create it).

But after running overnight, the loss seemed stuck ~14 for 16 hours after descending reliably, and I took a closer look. The data loading describes the MIDI format as being a piano roll format, which represents every possible instrument X every timestep, which seems like an absurdly enormous and sparse way to encode MIDI, and potentially disastrous for Reformer at a length of only 4096. 4096 is fine when you have a compact encoding like ABC, but piano roll doesn't seem compact at all.

To take a closer look, I added `print(output.size)` to `music.py`. Puzzlingly, this always prints out '3591' (for data length = 512, '28679' for l=4096, and '458759' for l=65536 before OOMing) while training. The conversion code is quite complex and I'm not sure what it's doing, but I also note that it avoids using the library's function for exactly this: https://craffel.github.io/pretty-midi/#pretty_midi.PrettyMIDI.get_piano_roll

~~~{.Python}
midi_to_roll(self.pathlist[idx], self.data_length + 1, self.augment) 
...
num = int(np.random.randint(0, len(input_list) - output_length + 1))
output = np.array(input_list[num : num + output_length], dtype=np.int64)
~~~

Is the data loading code 'solving' the sparsity by generating enormous sparse matrixes and then simply randomly sampling subsequences? It seems like it would be impossible to learn decent MIDI generation with a strategy like that. Literally most sequences for feasible lengths like l=4096 will simply be zeros and Reformer will never see two instruments simultaneously, making it impossible to condition instruments on each other. Even with a length like 65536 (which OOMs), I'm not sure piano roll formatting would fit enough into the window to allow useful generation, and it would be extremely slow to train on. There must be some better way of representing MIDI for Reformer's purposes.

**Comment by Rick-McCoy** [comment]
Hi, thanks for taking an interest in this project.

The MIDI representation used in this particular case is a version somewhat resembling what OpenAI did with their [MuseNet](https://openai.com/blog/musenet/), although heavily modified. In a nutshell, every timestep corresponds to a single note, represented by five variables: instrument, pitch, velocity, duration, and delta. Delta is the elapsed time between the current note and the previous note. Instrument, pitch, and velocity have size 129, 128, 128 vocabularies each, and duration & delta is size 16384, capable of representing up to 8 seconds of length with a resolution of 1 / 2048 seconds. Since a vocabulary size of 16384 is too large, these are split into two separate encodings of 128 each, making a total of seven encodings for each note. Then they are concatenated & linearly projected into a single encoding vector.

This method allows for extremely long MIDI files to be represented in a highly compact format, with even minutes-long files fitting in RAM. This doesn't use ```piano_roll```, although I've dabbled with them for my previous projects.

I also got the same problem as you described, with the loss getting stuck at an unsatisfying level and not converging further. I would be grateful for any sort of help and/or remarks.

**Comment by gwern** [comment]
I see. So it's not really a piano roll at all. How many tokens total do normal MIDI files encode to? I'm suspicious that such a format is a good idea...

One thing to try is ABC format. We have obtained excellent MIDI samples from training GPT-2-117M with 5k context windows on the ABC-encoded form of MIDIs, which is not quite lossless, as the average ABC is only ~10k BPEs total. So we know ABC-MIDI works with at least one Transformer architecture at relatively small windows. Could you try training on ABC text? The current codebase doesn't seem to support reading text, or else I would've tried that next.

**Comment by Rick-McCoy** [comment]
I've used the ABC format quite some time ago, in fact, while experimenting with Karpathy's excellent [char-rnn](https://karpathy.github.io/2015/05/21/rnn-effectiveness/). I quite liked my initial results, but unfortunately, multi-instrument support was not that effective, since separate instruments were listed completely detached from each other. For my goal was to create a universal-MIDI model (a quite ambitious one), I set out searching for another format.

I don't quite remember the number of tokens an average file consists of, but from what I recall, with length 4096 ~90% of my toy dataset [The Piano Dataset](http://piano-e-competition.com/) fit, and ~70% of the [Lakh midi dataset](https://colinraffel.com/projects/lmd/) fit. I thought that was enough.

**Comment by gwern** [comment]
char-RNN would never have worked because the hidden state fades out quickly and in practice it can hardly remember anything past a few hundred timesteps. The instruments are generated in big blocks and span thousands of timesteps, so instruments would just wind up being generated independently. However, that formatting works just fine with Transformers/attention and a window of 5k, and you can hear many multi-instrument pieces on my page.

If most of them are <4096, then the length should be OK I guess, and maybe the encoding is just too opaque to work with. I'd give ABC another try with Reformer rather than char-RNN.

**Comment by Rick-McCoy** [comment]
I mean, the char-rnn thing was a toy project back when I started ML. I understand that ABC Notation will work better with these architectures.

Thank you once again for taking interest in this implementation. Is it all right if I close this issue?

**Comment by gwern** [comment]
I'm not convinced the MIDI representation is a good one, but if you're not interested in testing out ABC, then I'll try some other implementation. The original Trax one I think supports text relatively easily, since [Tenoke got it running](https://svilentodorov.xyz/blog/reformer-99m/) although I'm wary of that tokenizing workflow...

**Comment by asigalov61** [comment]
@gwern 
@Rick-McCoy 

Hey guys!

Very interesting discussion and also very useful info. Thank you. I know that I am a year late to the party, but better late than never, right? ;)

Anyways, I specialize in Music AI and I wanted to invite you to check out my best work up-to-date that you may find useful for your projects.

https://github.com/asigalov61/Optimus-VIRTUOSO

This is GPT2+ GPT3 tweaks-based implementation that works just great with music. I have not had the same success with Reformer or any other SOTA transformers, as I did with GPT2. Plus, I use my own custom MIDI processors that are tuned to the max to be efficient while preserving musical elements such as timings and other stuff...

Any feedback will be appreciated and thank you again for the discussion and the codebase (I wanted to adopt some of it in my projects).

Sincerely,

Alex
